File: amlip_cpp/src/cpp/node/workload_distribution/AsyncComputingNode.cpp
// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @file AsyncComputingNode.cpp
 */

#include <cpp_utils/Log.hpp>
#include <cpp_utils/exception/InconsistencyException.hpp>

#include <amlip_cpp/node/workload_distribution/AsyncComputingNode.hpp>

#include <dds/multiservice/AsyncMultiServiceServer.hpp>
#include <dds/Participant.hpp>
#include <dds/network_utils/topic.hpp>

namespace eprosima {
namespace amlip {
namespace node {

struct TaskListenerCast : public dds::TaskListener<types::JobDataType, types::JobSolutionDataType>
{
    TaskListenerCast(
            const std::shared_ptr<JobReplier>& listener)
        : listener_(listener)
    {
    }

    types::JobSolutionDataType process_task (
            std::unique_ptr<types::JobDataType>&& task,
            const types::TaskId& task_id,
            const types::AmlipIdDataType& client_id,
            const types::AmlipIdDataType&) override
    {
        return listener_->process_job(*task, task_id, client_id);
    }

    std::shared_ptr<JobReplier> listener_;
};

AsyncComputingNode::AsyncComputingNode(
        const char* name,
        const std::shared_ptr<JobReplier>& listener,
        uint32_t domain_id)
    : ParentNode(name, types::NodeKind::computing, types::StateKind::stopped, domain_id)
    , job_server_(
        participant_->create_async_multiservice_server<types::JobDataType,
        types::JobSolutionDataType>(dds::utils::JOB_TOPIC_NAME))
    , listener_(listener)
{
    logInfo(AMLIPCPP_NODE_ASYNCCOMPUTING, "Created new Async Computing Node: " << *this << ".");
}

AsyncComputingNode::AsyncComputingNode(
        const char* name,
        const std::shared_ptr<JobReplier>& listener)
    : AsyncComputingNode(name, listener, dds::Participant::default_domain_id())
{
    // Do nothing
}

AsyncComputingNode::~AsyncComputingNode()
{
    if (current_state_ == types::StateKind::running)
    {
        stop();
    }
    logDebug(AMLIPCPP_NODE_ASYNCCOMPUTING, "Destroying Async Computing Node: " << *this << ".");
}

void AsyncComputingNode::run()
{
    if (current_state_ == types::StateKind::running)
    {
        throw utils::InconsistencyException(
                  STR_ENTRY << "Async Computing Node " << this << " is already running.");
    }
    else
    {
        logInfo(AMLIPCPP_NODE_ASYNCCOMPUTING, "Running Async Computing Node: " << *this << ".");
        job_server_->run(std::make_shared<TaskListenerCast>(listener_));
        change_status_(types::StateKind::running);
    }
}

void AsyncComputingNode::stop()
{
    if (current_state_ == types::StateKind::stopped)
    {
        throw utils::InconsistencyException(
                  STR_ENTRY << "Async Computing Node " << this << " is already stopped.");
    }
    else
    {
        logInfo(AMLIPCPP_NODE_ASYNCCOMPUTING, "Stopping Async Computing Node: " << *this << ".");
        job_server_->stop();
        change_status_(types::StateKind::stopped);
    }
}

} /* namespace node */
} /* namespace amlip */
} /* namespace eprosima */

File: amlip_py/amlip_py/node/AsyncComputingNode.py
# Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""AML-IP Async Main Node API specification."""

from amlip_py.types.AmlipIdDataType import AmlipIdDataType
from amlip_py.types.JobDataType import JobDataType
from amlip_py.types.JobSolutionDataType import JobSolutionDataType

from amlip_swig import AsyncComputingNode as cpp_AsyncComputingNode
from amlip_swig import JobReplier as cpp_JobReplier


class JobReplier(cpp_JobReplier):
    """
    TODO
    """

    def process_job(
            self,
            job,
            task_id,
            client_id):
        """
        Raise exception.
        Abstract method.
        This method should be reimplemented by child class.
        """
        raise NotImplementedError('JobReplier.process_job must be specialized before use.')


class JobReplierLambda(JobReplier):
    """
    Custom JobReplier supporting to create it with a lambda function.
    This object is created with a lambda function that is stored inside and used for every
    JobSolution message received.
    """

    def __init__(self, callback):
        """Construct new object with lambda as callback."""
        self.callback_ = callback
        super().__init__()

    def process_job(
            self,
            job,
            task_id,
            server_id):
        """Call internal lambda."""
        return self.callback_(job, task_id, server_id)


class AsyncComputingNode(cpp_AsyncComputingNode):
    """
    AML-IP Asynchronous Computing Node.
    TODO
    """

    def __init__(
            self,
            name: str,
            callback=None,
            listener: JobReplier = None):
        """
        Create a new Async Main Node with a given name.
        Parameters
        ----------
        name : str
            Name of the node.
        """

        if (listener and callback):
            raise ValueError(
                'AsyncComputingNode constructor expects a listener object or a callback, both given.')

        elif (not listener and not callback):
            raise ValueError(
                'AsyncComputingNode constructor expects a listener object or a callback, none given.')

        # Parent class constructor
        if listener:
            super().__init__(name, listener)
        elif callback:
            super().__init__(name, SolutionLambda(callback))
        else:
            raise ValueError(
                'This should not happen.')

    def run(
            self,
            ) -> None:
        """
        TODO
        """
        return cpp_AsyncComputingNode.run(self)

    def stop(
            self,
            ) -> None:
        """
        TODO
        """
        return cpp_AsyncComputingNode.stop(self)

    def get_id(
            self) -> AmlipIdDataType:
        """Get AMLIP id of the node."""
        return cpp_AsyncComputingNode.id(self)
